### PR TITLE
Preserve reference definitions around disabled regions

### PR DESCRIPTION
**Pull request by dahlia**
Fixes <https://github.com/dahlia/hongdown/issues/27>.

A reference definition is consumed by the parser into a map it does not expose, so it has no node in the tree. A disabled region was emitted by walking its nodes and copying each one's source span, which left nothing to copy a definition from. The definition vanished, the links inside the region pointed at nothing, and since the loss was stable across runs, `--check` never reported the file afterwards.

The issue reports this for `hongdown-disable`. It happens under every disable directive, for a second reason: a link that is copied never registers the definition it uses, so one living outside the region is dropped as unused. The badge in `test_directive_disable_enable` had been losing both its definitions for as long as that test existed, and the test passed because it only asserted that the badge text survived.

Copying the region instead of rebuilding it
-------------------------------------------

The region between the two directives is now taken from the source in one piece, the way `hongdown-disable-file` already took the rest of the file. Its reference definitions survive, and so do its footnote definitions, its indentation and the blank lines inside it. Only the blank lines against the directives themselves are normalized, to one on either side, which is what keeps a second run from adding another.

A footnote definition inside a region used to be relocated and reformatted. Now it stays where it was written, so the pass that collects footnotes has to leave it alone or it would be emitted twice, and a trailing comment inside a region is part of the copy rather than something to move below the definitions.

The blocks that `hongdown-disable-next-line` and `hongdown-disable-next-section` copy one at a time are now copied by their lines too. A block's span begins at its content, past the indentation and the markers of whatever holds it, so copying from there dropped them: an indented code block came out as a paragraph, and a definition the parser had taken out of a list item was dropped from the copy altogether. Both predate this branch.

Who owns a label
----------------

A copied link keeps whatever label the source gave it. A formatted link can be given a derived one, as in `[guide][guide 2]`. So when the two want the same label for different targets, the formatted link is the one that yields, which means the labels copied links use have to be claimed before anything is serialized.

A definition inside a copy holds its label whether or not a link resolves through it. It defines that label document-wide, and CommonMark keeps the first definition of one, so a later link taking it would find its own definition placed after the copy and quietly resolve to the copy's target instead. Its target is unknown in exactly the case where nothing resolves through it, so such a label cannot be shared, only avoided.

When a reserved definition falls due
------------------------------------

Registering a copied link's definition at the moment the copy was emitted let the copy decide where that definition landed, and what it landed among.

Where it landed was wrong because a copied link resolves only once its definition exists. Whether the registration happened at all therefore depended on whether an earlier run had already written it: on the first run the definition was placed by the document's own link, on the second by the copy, and one pass stopped settling the file. A reservation now falls due where the source puts the definition, the way a footnote's references follow their footnote. Since when it is made no longer decides where it lands, the three places that had been arranging this by hand became one pass made up front.

Finding the definitions at all
------------------------------

Comrak keeps its reference map private, so the definitions have to be recognized from the source. What makes that reliable is block context taken from the tree: the lines a leaf block occupies are excluded, container blocks are descended into so that the gaps between their children stay visible, and a line the parser kept out of every leaf block while looking like a definition is one. Prose that merely resembles a definition is part of its block and never reaches the test. This replaced a pile of heuristics about destinations and containers that kept getting the boundary cases wrong.

A paragraph complicates it by keeping the lines of a definition consumed at its head. Those lines have to stay visible to the scanner while the paragraph's own lines do not, so where the content starts is read two ways, by the line breaks the paragraph contains and by the definitions its head is made of, and a line is left visible only where both agree. Each reading misjudges what the other catches: a code span holding a newline for the first, an opening line that merely resembles a definition for the second. Comrak also reports an inline inside such a paragraph above its real line, by the height of the consumed head, so the reads that recover a copied link's label carry that offset.

Reading a definition is one function rather than two. It follows every part that may break across lines, and it treats container markers the way the parser does: a list marker before a definition is set aside, since `- [x]: done` is no task item, while below one a bullet opens an item and ends the definition, as does an ordered marker numbering from one. An ordered marker numbering from anywhere else goes on reading as part of the line above it.

Cost
----

The expensive half of that analysis, which is the definition scan and the walks that claim and reserve labels, runs only where a document has copied something. Membership of the copied lines is a line-indexed table rather than a list of spans to walk, because scanning a span per line cost time quadratic in the number of blocks. Locally, 64000 copied blocks went from 68 seconds to about 0.15, and a document with no directives at all is where it was at the branch point, around 0.12 seconds for 48000 blocks.

What this leaves alone
----------------------

Three defects turned up while testing that predate this branch and belong to the earlier reference work rather than to disabled regions. Two are filed: a definition sinking to a footnote's section when they share a label, https://github.com/dahlia/hongdown/issues/30, and labels differing only by non-ASCII edge whitespace being merged, https://github.com/dahlia/hongdown/issues/31. The third has no issue yet. Allocating a numbered label does not check whether that label already appears as unresolved reference syntax, so formatting can turn literal text like `See [guide 2].` into a link. This branch makes that one reachable in one more way, since a copied definition can push a formatted link into a numbered label, though the flaw is in the allocator.

Two repairs did land here that are not about disabled regions, because the code they were in is what produced the wrong answers above. A reference label holding a bracket escaped with a backslash, as in `[a\]b]`, was cut short at that bracket, which lost the link and left the definition unreadable on the next run. And `ensure_blank_line` no longer writes a separator into empty output, so a document whose first emitted thing is a definition does not start with blank lines.

Testing
-------

The branch adds 58 tests, many of which format twice and compare, since most of what went wrong here showed up as a file that `--write` wrote and `--check` then rejected. `cargo test` reports 775 unit tests and 79 integration tests. Running Hongdown over the repository's own Markdown produces no changes.

### CODE REVIEW

File: src/serializer/document.rs [hunk 1]
```diff
@@ -233,9 +385,686 @@ impl<'a> Serializer<'a> {
...
+                    underline_lines,
+                )
+            } else {
+                sourcepos.start.line
+            };
+            for line in start..=sourcepos.end.line {
+                if let Some(marked) = in_leaf_block.get_mut(line - 1) {
+                    *marked = true;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/document.rs]: Addressed in 67d875f. The marking now goes through mark_lines, which already holds that guard, so the max(1) lives in one place rather than two.

File: src/serializer/mod.rs [hunk 2]
```diff
@@ -40,6 +40,26 @@ fn strip_math_sentinels(mut output: String) -> String {
...
+            continue;
+        }
+        match character {
+            '\\' => escaped = true,
+            ']' => return Some(offset),
+            _ => {}
+        }
+    }
+    None
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/mod.rs]: Addressed in a3fbb62. The general scanner now lives beside the link reader that had the specialization, and the definition reader calls it; two readings of a label disagreeing about where it ends is a bug this branch has already had to repair twice.

File: src/serializer/tests.rs [hunk 3]
```diff
@@ -1280,6 +1280,896 @@ fn test_directive_disable_enable() {
...
+fn test_directive_disable_reservation_precedes_a_setext_heading() {
+    // The definition sits on the line a setext heading reports as its own
+    // start, but it precedes the heading, so it belongs above the section.
+    let input = "<!-- hongdown-disable -->\n\nRaw [a].\n\n<!-- hongdown-enable -->\n\n[a]: https://example.com/first\nSection\n-------\n\nBody.\n";
+    let result = parse_and_serialize_with_source(input);
+    assert!(
+        result.find("[a]: https://example.com/first") < result.find("Section\n-------"),
+        "the definition must come before the section it precedes, got:\n{}",
+        result
+    );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/tests.rs:1715]: Addressed in ee7deae, and you are right that it was exactly the guarded regression that slipped through: removing the reservation left every one of those tests passing. All nine sites now take each position with a message of its own before comparing.

File: src/serializer/document.rs [hunk 4]
```diff
@@ -91,59 +194,94 @@ impl<'a> Serializer<'a> {
...
+                                .sourcepos
+                                .start
+                                .line
+                                .saturating_sub(1),
+                            None => self.source_lines.len(),
+                        };
+                        let region = self
+                            .extract_source_lines(region_start, region_end)
+                            .map(|source| Self::trim_blank_lines(&source))
+                            .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/document.rs:235]: Addressed in f652733: a "Disabled regions" section covering what is copied as written, the one blank line that is normalized beside a directive comment, and the definitions a copied link needs, tied back to the rule on distinct labels for distinct targets. Both examples in it are checked against the formatter and round-trip unchanged.

File: src/serializer/document.rs [hunk 5]
```diff
@@ -91,59 +194,94 @@ impl<'a> Serializer<'a> {
...
+                        if region.is_empty() {
+                            // Nothing to copy, either because the region is
+                            // empty or because the source is unavailable; fall
+                            // back to emitting its blocks one by one.
+                            self.skip_mode = FormatSkipMode::Disabled;
+                        } else {
+                            self.output.push('\n');
+                            self.output.push_str(&region);
                             self.output.push('\n');
+                            skip_until = region_last;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/document.rs:272]: Addressed in f948783. Reproduced first: with sentence case on, a heading after the region came out as "Using swift later". The nouns are now taken up as the region is copied; the comment itself was already part of the copy, so only the effect was missing. Regression test in tests/integration.rs.

File: src/serializer/document.rs [hunk 6]
```diff
@@ -91,59 +194,103 @@ impl<'a> Serializer<'a> {
...
+                        } else {
+                            // A directive naming nouns names them for the
+                            // headings after the region as much as for the ones
+                            // before it, which is what it did when the region
+                            // was emitted a node at a time.  Its comment is
+                            // part of the copy; only its effect is left to
+                            // carry out.
+                            for skipped in &children[i + 1..region_last] {
+                                self.apply_noun_directive(skipped);
+                            }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply src/serializer/document.rs:265]: Addressed in 168bb6d. Confirmed against the branch point first: with a list after the enable, it used to keep its own marker and now had it rewritten, so the directive really had stopped reaching past the region. The region is scanned for one before it is copied; where it holds one, the copy runs to the last line and nothing follows it.
